archive_pipeline/archive.py

```diff
@@ -0,0 +1,147 @@
+
+
+def get_deleted_rows(conn, delete_timestamp: str) -> list[tuple]:  # pragma: no cover
+    """Returns the rows that were recorded more than a day ago."""
+    with conn.cursor() as cur:
+        print("Fetching data to archive from RDS.")
+        cur.execute("""
+            SELECT p.plant_entry_id AS entry_id,
+                s.scientific_name AS species,
+                p.temperature,
+                p.soil_moisture,
+                p.humidity,
+                p.last_watered AS last_watered,
+                p.recording_taken AS recording_taken,
+                sun.s_description AS sunlight,
+                b.b_name AS botanist_name,
+                c.cycle_name AS cycle
+            FROM plant p
+            LEFT JOIN species s ON p.species_id = s.species_id
+            LEFT JOIN sunlight sun ON p.sunlight_id = sun.sunlight_id
+            LEFT JOIN botanist b ON p.botanist_id = b.botanist_id
+            LEFT JOIN cycle c ON p.cycle_id = c.cycle_id
+            WHERE recording_taken < %s;""", (delete_timestamp,))
+        deleted_rows = cur.fetchall()
+        conn.commit()
+        cur.close()
+        print("Archive data fetched from RDS.")
+        return deleted_rows
+
+
+def select_and_delete_from_db(conn, delete_timestamp: str, configuration: dict) -> None:
+    """Removes rows from the plant table that are older than a day.
+
+    SELECT is performed to maintain consistency with live databatase interactions in the dashboard.
+    Function creates the .csv and uploads it to S3 before deleting any rows.
+    """
+    rows_to_delete = get_deleted_rows(conn, delete_timestamp)
+
+    deleted_rows_df = create_deleted_rows_dataframe(rows_to_delete)
+    archived_csv_filename = create_csv_filename()
+    create_archived_csv_file(deleted_rows_df, archived_csv_filename)
+    upload_csv_to_s3(archived_csv_filename, configuration)
+
+    deleted_rows = delete_old_rows(conn, delete_timestamp)
+
+    if len(rows_to_delete) != len(deleted_rows):
+        print("Inconsistency between rows being deleted and rows being archived.")
+
+
+def create_deleted_rows_dataframe(deleted_rows: list[tuple]) -> pd.DataFrame:
+    """Returns a single DataFrame containing all the rows of the deleted data."""
+    archive_df = pd.DataFrame(deleted_rows, columns=CSV_COLUMNS)
+    if archive_df.empty:
+        raise ValueError("Empty DataFrame.")
+    return archive_df
+
+
+def create_csv_filename() -> str:
+    """Creates a filename for a .csv file with the previous day's date."""
+    today_date = date.today()
+    yesterday_date = today_date - timedelta(days=1)
+    yesterday_date_str = datetime.strftime(yesterday_date, "%Y_%m_%d")
+    return "archived_" + yesterday_date_str + ".csv"
+
+
+def create_archived_csv_file(archived_df: pd.DataFrame, csv_filename: str) -> None:
+    """Creates a .csv file from a Pandas dataframe."""
+    if os.path.exists(csv_filename):
+        print(f"A file already exists locally with the name {csv_filename}. " +
```

Comment:
We will be getting data every minute, the filename if only made up by date will surely get re-written for every minute of the day? Are we trying to keep all of one day's data in one file -- in which case we will have to append to an existing CSV where the date is the same?

Reply:
This pipeline will only run once every 24 hours. So everything will go into a single .csv for that 24 hours.

Reply:
Wait so who is doing the lambda to run the extract every minute?